core/src/main/java/com/linecorp/armeria/internal/common/stream/HttpDataDeframer.java

```diff
@@ -0,0 +1,417 @@
+                length += buf.capacity();
+            }
+            return length;
+        }
+
+        /**
+         * Returns the number of readable bytes which is equal to {@code (length() - readerIndex()}.
+         */
+        public int readableBytes() {
+            if (queue.isEmpty()) {
+                return 0;
+            }
+
+            int readableBytes = 0;
+            for (ByteBuf buf : queue) {
+                readableBytes += buf.readableBytes();
+            }
+            return readableBytes;
+        }
+
+        /**
+         * Returns the {@code readerIndex} of this {@link HttpDataReader}.
+         * the {@code readerIndex} will be set to {@code 0} after {@link #discardReadBytes()} is called.
+         */
+        public int readerIndex() {
+            if (queue.isEmpty()) {
+                return 0;
+            }
+
+            int value = 0;
+            for (ByteBuf buf : queue) {
+                final int readerIndex = buf.readerIndex();
+                if (readerIndex > 0) {
+                    value += readerIndex;
+                } else {
+                    break;
+                }
+            }
+            return value;
+        }
+
+        /**
+         * Sets the {@code readerIndex} of this {@link HttpDataReader}.
+         */
+        public void readerIndex(int readerIndex) {
+            int remained = readerIndex;
+            for (ByteBuf buf : queue) {
+                final int writerIndex = buf.writerIndex();
+                if (writerIndex >= remained) {
+                    buf.readerIndex(remained);
+                    return;
+                } else {
+                    remained -= writerIndex;
+                }
+            }
+        }
+
+        /**
+         * Returns a byte at the current {@code readerIndex} and increases the {@code readerIndex} by {@code 1}.
+         */
+        public byte readByte() {
+            for (ByteBuf buf : queue) {
+                if (!buf.isReadable()) {
+                    continue;
+                }
+                return buf.readByte();
+            }
+            throw new IndexOutOfBoundsException("readerIndex: " + readerIndex() +
+                                                " (expected: 0 <= readerIndex < length(" + length() + "))");
+        }
+
+        /**
+         * Returns a unsigned byte at the current {@code readerIndex} and increases the {@code readerIndex}
+         * by {@code 1}.
+         */
+        public short readUnsignedByte() {
+            return (short) (readByte() & 0xFF);
+        }
+
+        /**
+         * Returns a 32-bit integer at the current {@code readerIndex} and increases the {@code readerIndex}
+         * by {@code 4}.
+         */
+        public int readInt() {
+            // fast path
+            final ByteBuf firstBuf = queue.peek();
+            if (firstBuf.readableBytes() >= 4) {
+                return firstBuf.readInt();
+            }
+
+            // slow path
```

Comment:
Slow path needs to be extracted into a separate method, so that the fast path part can be inlined. If you put both slow and fast path into a single method, it will never be inlined since it's too big.

Reply:
It a good tip! 👍